front/pages/api/v1/w/[wId]/spaces/[spaceId]/data_sources/index.ts

```diff
@@ -58,6 +58,16 @@ async function handler(
 ): Promise<void> {
   const dataSources = await DataSourceResource.listBySpace(auth, space);
 
+  if (!space.canList(auth)) {
```

Comment:
this logic is already performed via the withResourceFromRoute wrapper; IMO we shouldn't duplicate the check

Reply:
We want to enforce the invariant that all endpoints should be excplicit about the authorization they require + rely on wrappers to provide defense in depth + rely on lower level to provide additional defense in depth.

I think authentication is important enough to make it an invariant that all endpoints have to be explicit about it even if we rely on wrappers to provide defense in depth. The argument is that relying on wrappers can lead to false expectations or even mis-interpretation as their implementation evolves.

Reply:
> generalize withResourceFromRoute so it can take multiple resources, and require the caller to specify the permissions to check on those, e.g. withResource**s**FetchingFromRoute(handler, { "space": ["read", "write"], "datasource": ["read"} }

The wrapper described in my comment this PR, that i'm quoting above for reference, is quite explicit about permissions, and has the advantage of requiring it explicitly when you use it to fetch a resource. So if you want resource X in a wrapper, you have to specify its permissions. 

IMO it's substantially better than wishing for coders to remember that we want to be explicit and copy-paste correctly every time.

In this PR, the code in this `if` block is really a copy of the code in the wrapper, with the missing case of the protection against conversations space access --- which strengthen the argument in favor of the wrapper: it's good that the logic is in a single place so that we avoid divergence.

This duplication is misleading. Any newcomer looking at this will think: "why do they do the exact same check and return the same apiError in the handler and in the wrapper?"

If we really wished to enforce this invariant, then why not do it for auth too? the checks are in the `withAuthentication` wrappers, but authentication is is important enough to make it an invariant that all endpoints have to be explicit about it => in that case we would reintroduce explicit auth checks in the handlers, no?

Glad to discuss IRL if needed :)